Port `nll_loss_backward` to structured

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62144

Differential Revision: [D29945279](https://our.internmc.facebook.com/intern/diff/D29945279)

Commit:
Port `nll_loss_backward` to structured

[ghstack-poisoned]

Commit:
Update on "Port `nll_loss_backward` to structured"




[ghstack-poisoned]

Commit:
Update on "Port `nll_loss_backward` to structured"




[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29945279).

ezyang commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -51,6 +53,56 @@ TORCH_META_FUNC(nll_loss_forward)
+        grad_output.sizes());
+  }
+
+  TensorOptions options(self.options());
```
TensorOptions is entirely functional, so you can do this fluently (`self.options().memory_format(...)`)

Commit:
Update on "Port `nll_loss_backward` to structured"


Differential Revision: [D29945279](https://our.internmc.facebook.com/intern/diff/D29945279)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29945279).

ezyang commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -52,6 +54,55 @@ TORCH_META_FUNC(nll_loss_forward)
+      total_weight.numel(),
+      " elements)");
+
+  auto weight = weight_opt.getTensorRef();
```
`const auto& weight`

ezyang reviewed (approved):
excellent!

Commit:
Update on "Port `nll_loss_backward` to structured"


Differential Revision: [D29945279](https://our.internmc.facebook.com/intern/diff/D29945279)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29945279).
